File: src/services/crawlerService.js
'use strict';

class CrawlerService {
  constructor(sources) {
    if (!sources || sources.length === 0) {
      throw new Error('CrawlerService requires at least one JobSource');
    }
    this._sources = sources;
  }

  async search(role, location, pages = 2) {
    const seen = new Set();
    const results = [];

    for (const source of this._sources) {
      console.log(`\n[Crawler] ${source.sourceName} → "${role}" in "${location}"`);
      try {
        const jobs = await source.search(role, location, pages);
        let added = 0;
        for (const job of jobs) {
          if (!seen.has(job.id)) {
            seen.add(job.id);
            results.push(job);
            added += 1;
          }
        }
        console.log(`[Crawler] ${source.sourceName}: +${added} unique jobs (${results.length} total)`);
      } catch (e) {
        console.log(`[Crawler] ${source.sourceName} error: ${e.message}`);
      }
    }

    return results;
  }
}

module.exports = { CrawlerService };

File: python/src/services/crawler_service.py
from ..interfaces.job_source import JobSource
from ..models import Job


class CrawlerService:
    def __init__(self, sources: list[JobSource]) -> None:
        if not sources:
            raise ValueError("CrawlerService requires at least one JobSource")
        self._sources = sources

    async def search(self, role: str, location: str, pages: int = 2) -> list[Job]:
        seen: set[str] = set()
        results: list[Job] = []

        for source in self._sources:
            print(f"\n[Crawler] {source.source_name} → \"{role}\" in \"{location}\"")
            try:
                jobs = await source.search(role, location, pages)
                added = 0
                for job in jobs:
                    if job.id not in seen:
                        seen.add(job.id)
                        results.append(job)
                        added += 1
                print(f"[Crawler] {source.source_name}: +{added} unique jobs ({len(results)} total)")
            except Exception as e:
                print(f"[Crawler] {source.source_name} error: {e}")

        return results
